source/adapters/hip/memory.cpp

```diff
@@ -282,16 +282,34 @@ UR_APIEXPORT ur_result_t UR_APICALL urMemGetInfo(ur_mem_handle_t hMemory,
 /// \param[out] phNativeMem Set to the native handle of the UR mem object.
 ///
 /// \return UR_RESULT_SUCCESS
-UR_APIEXPORT ur_result_t UR_APICALL urMemGetNativeHandle(ur_mem_handle_t,
-                                                         ur_native_handle_t *) {
-  // FIXME: there is no good way of doing this with a multi device context.
-  // If we return a single pointer, how would we know which device's allocation
-  // it should be?
-  // If we return a vector of pointers, this is OK for read only access but if
-  // we write to a buffer, how would we know which one had been written to?
-  // Should unused allocations be updated afterwards? We have no way of knowing
-  // any of these things in the current API design.
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+UR_APIEXPORT ur_result_t UR_APICALL
+urMemGetNativeHandle(ur_mem_handle_t hMem, ur_device_handle_t Device,
+                     ur_native_handle_t *phNativeMem) {
+  if (!Device)
+    return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+#if defined(__HIP_PLATFORM_NVIDIA__)
+  if (sizeof(BufferMem::native_type) > sizeof(ur_native_handle_t)) {
+    // Check that all the upper bits that cannot be represented by
+    // ur_native_handle_t are empty.
+    // NOTE: The following shift might trigger a warning, but the check in the
+    // if above makes sure that this does not underflow.
+    BufferMem::native_type UpperBits =
+        std::get<BufferMem>(hMem->Mem).getPtr(Device) >>
+        (sizeof(ur_native_handle_t) * CHAR_BIT);
+    if (UpperBits) {
+      // Return an error if any of the remaining bits is non-zero.
+      return UR_RESULT_ERROR_INVALID_MEM_OBJECT;
```

Comment:
It may be good to extend the UR API declaration's documentation of this API with this return code (`UR_RESULT_ERROR_INVALID_MEM_OBJECT`) as it was not part of the original version. However, since it is `__HIP_PLATFORM_NVIDIA__`-specific it may or may not be useful readers if we chop the HIP on NVIDIA support completely but it is still a valid case today.

I am not adding this as a blocker to the PR because it is so niche due to not being related to AMD hip target devices but I think because we currently still care to support this one case we ought to document all possible result codes accordingly.

Reply:
Yeah I'm not sure about this. At the moment there isn't a precedent for explaining return codes specific to the HIP and CUDA adapter, much less for the HIP Nvidia adapter, so I would be inclined to let this slide.

Also bear in mind that this code is merely copy pasted from what was there before the multi dev ctx PR https://github.com/oneapi-src/unified-runtime/pull/999/files#diff-cdda61bf8ead17f141255a7bceb4231d050653676e565fde5aa6449a30a428b2L321

Reply:
Yeah, it's why I mentioned I wasn't too sure either. I'll leave it to your best judgement then. Thanks!

Reply:
Thanks @GeorgeWeb